docs/contributing/Android-API-And-Related-Versions.md

```diff
@@ -0,0 +1,115 @@
+# Android Api and Related versions
+
+Picking the full suite of android versions can be difficult if you are not a full time android developer. 
+This document is for contributors to understand what versions of common android tooling should be used in the flutter/flutter codebase. 
+
+## Overview
+
+**The most important thing is for any test, project, or app to have compatible versions. **
+
+If versions need to be different then a comment explaining why and what is being evaluated should be included. 
+All chosen versions should pass the minimum versions checks defined in [DependencyVersionChecker](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/gradle/src/main/kotlin/DependencyVersionChecker.kt) 
+
+If the versions chosen are not known by [gradle_utils.dart](https://github.com/flutter/flutter/blob/a16c447abcb695b4ca907d59e66dc87f4f7178d3/packages/flutter_tools/lib/src/android/gradle_utils.dart#L63) then update gradle_utils.dart. 
+
+## Specifics
+
+- "compileSdk" version should be `flutter.compileSdkVersion` or if a pure android project the highest number that has proven stable and is available in CIPD. 
+- "targetSdk" takes human level effort to update. This is by design of the AGP team. The version should be `flutter.targetSdkVersion` or the highest number that has proven stable and is available in cipd. This number may lag behind compile sdk because of the effort to update.
+- "AGP" version should be the version set in flutter templates or newer. If the version is intentionally different there should be a comment explaining why.
+- "gradle" version should be the version set in flutter templates or newere. If the version is intentionally different there should be a comment explaining why.
+- "kotlin" version should be the version set in flutter templates or newere. If the version is intentionally different there should be a comment explaining why.
+
+### compileSdk 
+
+Must be higher then or equal to `targetSdk`. 
+Should be `compileSdk` instead of `compileSdkVersion`. 
+Should be the max value available in CIPD if not `flutter.compileSdkVersion`
```

Comment:
(Non-blocking) Could we encode this somehow, i.e. `flutter.compileSdkMaximum`?

Reply:
I wouldnt want to for a couple of reasons first the values in `flutter.` are public and we dont want the code to have another value that users might get confused by.  Second as a practice the max is a value if you dont have access to flutter.compileSdkVersion for example in an add to app code sample. 

I think longer term @gmackall has a plan to divorce the flutter. values from the flutter gradle plugin as something like a flutter values gradle plugin. That will let us use flutter. variables in non flutter apps like add to app.